### ISSUE TITLE
Decimals are json-dumped as strings

### ISSUE THREAD

**Issue by Martin19037** [pydantic/pydantic#7457]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Is it intended that Decimals come out as strings when dumped to json? I could not find any documentation for this behavior, so I assume this is a bug.

### Example Code

```Python
from decimal import Decimal
from pydantic import BaseModel, RootModel, TypeAdapter

class TestModel(BaseModel):
    int_field: int
    float_field: float
    decimal_field: Decimal

test_model = TestModel(
    int_field=1,
    float_field=1.1,
    decimal_field=Decimal('1.2')
)
json_dump_1 = test_model.model_dump(mode='json')  # {'decimal_field': '1.2', 'float_field': 1.1, 'int_field': 1}
json_dump_2 = test_model.model_dump_json()  # '{"int_field":1,"float_field":1.1,"decimal_field":"1.2"}'

class DecimalRootModel(RootModel[Decimal]):
    pass

root_model = DecimalRootModel(Decimal('1.2'))
json_dump_3 = root_model.model_dump(mode='json')  # '1.2'
json_dump_4 = root_model.model_dump_json()  # '"1.2"'

decimal_adapter = TypeAdapter(Decimal)
json_dump_5 = decimal_adapter.dump_python(Decimal('1.2'), mode='json')  # '1.2'
json_dump_6 = decimal_adapter.dump_json(Decimal('1.2'))  # b'"1.2"'
```

### Python, Pydantic & OS Version

```Text
             pydantic version: 2.3.0
        pydantic-core version: 2.6.3
          pydantic-core build: profile=release pgo=false
                 install path: C:\***\venv_311\Lib\site-packages\pydantic
               python version: 3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)]
                     platform: Windows-10-10.0.19045-SP0
     optional deps. installed: ['email-validator', 'typing-extensions']
```

**Comment by sydney-runkle** [comment]
Hi @Martin19037, thanks for your question!

This is the expected behavior of `Decimal` serialization. The reason for this is that this allows values such as `Decimal('NaN')` and `Decimal('Infinity')` to be serialized correctly. We have a few [tests](https://github.com/pydantic/pydantic-core/blob/f6b14cce22b1539727ed0e0a47d2c680c79c1692/tests/serializers/test_decimal.py#L46) in the `pydantic-core` repo showing this behavior.

If you're wanting to serialize your decimal values as floats, you could take the following approach:
```py
from decimal import Decimal
from typing import Annotated

from pydantic import BaseModel, PlainSerializer, TypeAdapter

class TestModel(BaseModel):
    int_field: int
    float_field: float
    decimal_field: Annotated[Decimal, PlainSerializer(lambda x: float(x), return_type=float, when_used='json')]

test_model = TestModel(
    int_field=1,
    float_field=1.1,
    decimal_field=Decimal('1.2')
)

print(test_model.model_dump())
#> {'int_field': 1, 'float_field': 1.1, 'decimal_field': Decimal('1.2')}
print(test_model.model_dump(mode='json'))
#> {'int_field': 1, 'float_field': 1.1, 'decimal_field': 1.2}
print(test_model.model_dump_json())
#> {"int_field":1,"float_field":1.1,"decimal_field":1.2}
```

Just be aware that you might run into issues if you used the following:

```py
from decimal import Decimal
from typing import Annotated

from pydantic import BaseModel, ConfigDict, PlainSerializer, TypeAdapter

class TestModel(BaseModel):
    model_config = ConfigDict(allow_inf_nan=True)
    
    int_field: int
    float_field: float
    decimal_field: Annotated[Decimal, PlainSerializer(lambda x: float(x), return_type=float, when_used='json')]

test_model = TestModel(
    int_field=1,
    float_field=1.1,
    decimal_field=Decimal('Infinity')
)

print(test_model.model_dump_json())
#> {"int_field":1,"float_field":1.1,"decimal_field":null}
```

**Comment by Martin19037** [comment]
Hi @sydney-runkle, thank you for the clarification!

This should be added to the documentation, as it is not what you would intuitively expect.

I understand that for special decimal values like `NaN`, `Infinity` and `-Infinity`, there is no json equivalent, and so dumping as a string is the next best thing. However, for the regular numeric domain, I cannot really agree with this argument.

On the other hand, if you compare it with floats, where `nan`, `inf` and `-inf` also exist, regular values are rendered as json numbers, and special values produce `null`.

If this is some technical or intentional design restriction, I am fine with it as long as it is well documented.

**Comment by sydney-runkle** [comment]
Hi @Martin19037,

Sure thing. I'll open a PR shortly to reflect this serialization pattern in the documentation.

As a side note -- this question has been raised before -- see #7120 regarding the intentionality of this serialization decision.

**Comment by matt7aylor** [comment]
Hi @sydney-runkle ,

Thanks for the documentation on these changes, this looks promising for allowing me to keep my numbers as numbers rather than an awkward change to strings. However, one issue with the above method is that the model json schema will still include `string` as a possible type, as for your example above `test_model.model_json_schema()` will produce:
```json
{
  "properties": {
    "int_field": {
      "title": "Int Field",
      "type": "integer"
    },
    "float_field": {
      "title": "Float Field",
      "type": "number"
    },
    "decimal_field": {
      "anyOf": [
        {
          "type": "number"
        },
        {
          "type": "string"
        }
      ],
      "title": "Decimal Field"
    }
  },
  "required": [
    "int_field",
    "float_field",
    "decimal_field"
  ],
  "title": "TestModel",
  "type": "object"
}

```
whereas a comparable model for pydantic v1 `test_model.schema_json()` will produce:
```json
{
  "title": "TestModel",
  "type": "object",
  "properties": {
    "int_field": {
      "title": "Int Field",
      "type": "integer"
    },
    "float_field": {
      "title": "Float Field",
      "type": "number"
    },
    "decimal_field": {
      "title": "Decimal Field",
      "type": "number"
    }
  },
  "required": [
    "int_field",
    "float_field",
    "decimal_field"
  ]
}
```

So, for example, my FastAPI generated schema still reports that the `Decimal` output could be serialized as a string (even with your method to serialize to a number), which kind of defeats the point of trying to make them `number` since I will still have to change the API clients to handle string so that they match the spec (or rewrite the whole API to avoid using Decimal). Do you know of any way to recover the `number` only functionality from pydantic v1 including the output json schema?

**Comment by sydney-runkle** [comment]
Hi @matt7aylor,

Thanks for your question and the detailed description 🤩 .

You can use the [`mode`](https://docs.pydantic.dev/latest/api/base_model/#pydantic.main.BaseModel.model_json_schema) flag to control the schema. So if you use `test_model.model_json_schema(mode='serialization')` then you'll get the following output:

```
{
    'properties': {
        'decimal_field': {
            'title': 'Decimal Field',
            'type': 'number',
        },
        'float_field': {
            'title': 'Float Field',
            'type': 'number',
        },
        'int_field': {
            'title': 'Int Field',
            'type': 'integer',
        },
    },
    'required': [
        'int_field',
        'float_field',
        'decimal_field',
    ],
    'title': 'TestModel',
    'type': 'object',
}
```

This output reflects the fact that `PlainSerializer` ensures that the `decimal_field` is serialized as a float type (when we're using `model_dump_json`).

**Comment by matt7aylor** [comment]
Hi @sydney-runkle,

Thanks for this, that does make sense, although I'm not really in control of how the json schema is obtained (tbh I'm just assuming it is like that), since that is handled by FastAPI (perhaps this would be a sensible thing for FastAPI to implement in schema generation, I'm not sure). What I would ideally like is a type I can use to preserve the Decimal as number functionality so that I don't have to rewrite my API to avoid Decimals (when updating to pydantic v2).

I have managed to do the following (there is an extra fiddle to preserve the integer decimals come out as `1` rather than `1.0` to match the behaviour of FastAPI before pydantic v2). This seems to do what I want but without me knowing much pydantic it feels potentially a bit crude/hacky (it also means models using this type would also accept floats rather than strict Decimals, which is fine in my case but maybe not ideal in general):

```python
def decimal_encoder(dec_value: Decimal) -> int | float:
    """
    Encodes a Decimal as int of there's no exponent, otherwise float
    """
    if dec_value.as_tuple().exponent >= 0:
        return int(dec_value)
    else:
        return float(dec_value)


NumberDecimal = (
    SkipJsonSchema[
        Annotated[Decimal, PlainSerializer(decimal_encoder, return_type=(int | float), when_used="json")]
    ]
    | float
)

class NumberModel(BaseModel):
    num: NumberDecimal

m = NumberModel(num=Decimal('1'))
m2 = NumberModel(num=Decimal('2.1'))

m.model_dump_json()
#> '{"num":1}'

m2.model_dump_json()
#> '{"num":2.1}'

m.model_json_schema()
#> {'properties': {'num': {'title': 'Num', 'type': 'number'}},
# 'required': ['num'],
# 'title': 'NumberModel',
# 'type': 'object'}
```

Do you have anything to suggest regarding potential issues or better ways of doing something along these lines? Or perhaps should I instead be looking deeper into the FastAPI schema generation to suggest a change there.

**Comment by samuelcolvin** [comment]
In v1 decimals we're dumped as floats and people complained a lot, said it was obviously the wrong thing due to the risk of loosing precision. I'm sure you can find the issues if you look.

Hence in v2 we changed it, to make people happy.

There are always trade offs, and something what is "obviously right"  to one person is "obviously wrong" to another.

**Comment by samuelcolvin** [comment]
See #1511.

**Comment by matt7aylor** [comment]
@samuelcolvin Thanks, I was aware of that and was/am not disputing the choice (counter-intuitive as it might seem). I'm merely trying to help with workarounds for those, such as myself, who have existing implementations built on the previous 'Decimal as number' method that are now broken with this new 'Decimal as string' approach.

**Comment by samuelcolvin** [comment]
You're code above looks good to me, my only concern would be whether you really want the `| float`, but I guess you have a reason for that.

**Comment by samuelcolvin** [comment]
@Kludex @sydney-runkle we should add this example to the example section of our docs.

**Comment by matt7aylor** [comment]
> You're code above looks good to me, my only concern would be whether you really want the `| float`, but I guess you have a reason for that.

Thanks for looking at the code. The `| float` is there so that the above `NumberDecimal` type will be able to produce a json schema, without it you will get an `AttributeError: 'NumberModel' object has no attribute 'model_schema_json'`. I agree the method is a bit crude though.

Perhaps the right approach is to see if FastAPI can incorporate the `.model_json_schema(mode='serialization')` mode change for schema generation. This way the simpler `PlainSerializer` only method that @sydney-runkle has already documented could potentially be used.

**Comment by matt7aylor** [comment]
Actually, and apologies on my part for not spotting this earlier, FastAPI recently has implemented a change to use the serialization mode (unfortunately I was testing with a feature branch for another issue and didn't have this change). So with the documented `PlainSerializer` method as above this should solve the issues I am encountering.

Thanks for your help and support on this. Hopefully this thread will be useful to others who encounter the same problems.

**Comment by sydney-runkle** [comment]
Hi @matt7aylor,

Thanks for following up. I'm glad the FastAPI change has resolved the issues you were encountering. I'm sure this discussion will be helpful to others with similar questions in the future 😄.

@samuelcolvin the [PR used to close this issue](https://github.com/pydantic/pydantic/pull/7465) adds the above example to the documentation 👍. Let me know if you'd like me to add more info!

**Comment by mortalisk** [comment]
Json has arbitrary precicion numbers, I don't get why this issue was worked around instead of fixing the serialization.

**Comment by mswezey23** [comment]
Glad I found this thread. Updating from v1 to v2 had some unit test snapshots changed that weren't expected. Definitely a breaking change that should have been called out in the docs sooner!

**Comment by brycedrennan** [comment]
> Json has arbitrary precicion numbers, I don't get why this issue was worked around instead of fixing the serialization

Yes. I'm trying to understand as well. JSON is ultimately a string so can we just output the characters that come from `str(Decimal("0.333333333"))` but without quotation marks? 

the key is to not needlessly convert to float since that will lose precision.

**Comment by samuelcolvin** [comment]
@brycedrennan yes but the library/language parsing the JSON will almost certainly use an f32 or f64 to represent floats from the JSON, so data is still lost in the most common case.

**Comment by brycedrennan** [comment]
Love pydantic. Thanks for responding.   Playing devils advocate here:

- Whether a consumer of my api cares about the extra precision is up to them, but I will have done my part.
- You can also imagine scenarios where floats are *usually* fine but the numbers represent values too important to allow float imprecision to be a potential issue (like money values).  Strings work fine unless you are parsing json sent to you with high-precision "non-strings" or a third party API is expecting you to post accurate non-string numbers. 
- I know some language somewhere is capable of making these high-precision json files because I have encountered json files I needed to consume that have higher precision than float.   The current system is unable to parse high precision numbers from json even for pydantic fields that are typed as `Decimal` (unless they are strings).   This caused me difficulty because it made it impossible to edit a completely unrelated field without causing a bunch of unrelated modifications to the numeric values. 

Example:

**Comment by Woody1193** [comment]
To me this would boil down to determining whether a field should be quoted or not in the resulting JSON. This doesn't hold true just for `decimal.Decimal` but also for Boolean types as those could either be represented as `"true"` or `true`. We could skirt this issue nicely by adding an option to the serializer to define the output field type as "Numeric", "Boolean" or "String" with that defining how the quotes should work.

**Comment by pn** [comment]
For me `mode='serialization'` was not working somehow when using latest FastAPI (0.112.2) and produced `string` in JSON schema. I have put
```python
model_config = ConfigDict(json_schema_mode_override='serialization')
```
as a workaround on every model with a Decimal with PlainSerializer field (as defined in https://github.com/pydantic/pydantic/issues/7457#issuecomment-1721431028). I hope this helps someone.

**Comment by sydney-runkle** [comment]
@pn, feel free to open a new issue re JSON schema generation if you'd like - sounds like we might not have the correct behavior by default there.

**Comment by sdasun** [comment]
I've created this new type so, I can reuse:

```python
from decimal import Decimal
from typing import Annotated
from pydantic import PlainSerializer

FloatDecimal = Annotated[
    Decimal,
    PlainSerializer(lambda x: float(x), return_type=float, when_used='json')
]

```

**Comment by samuelcolvin** [comment]
@Viicos we should provide a config option to serialize decimals to JSON directly within pydantic-core. Not just for convenience, but also because there's currently no way to serialize a decimal as a float without contrainting precision to 64 bits. E.g. @sdasun's solution involves converting the decimal to a (64bit) float.

BTW, you can slightly simplify @sdasun's solution to

```py
FloatDecimal = Annotated[
    Decimal,
    PlainSerializer(float, return_type=float, when_used='json')
]
```

**Comment by gdub** [comment]
Stumbled across this issue while dealing with v1 -> v2 upgrades.

Would love to see option to allow direct serialization/deserialization between Python Decimal type and JSON number type. I have use case dealing with values where I don't want to lose precision nor significant trailing zeros with conversion through float.

With Pydantic v1, I was able to make use of the `json_loads`/`json_dumps` config options to use [rapidjson](https://github.com/python-rapidjson/python-rapidjson). rapidjson has a few different number-handling modes, including [`NM_DECIMAL`](https://python-rapidjson.readthedocs.io/en/latest/api.html#rapidjson.NM_DECIMAL) mode that serializes all Python `Decimal` values to JSON `number` and deserializes all JSON `integer` and `number` values to Python `Decimal` without loss. For example:

```python
from decimal import Decimal
from pydantic.v1 import BaseModel
import rapidjson


def rapidjson_dumps(obj, *args, **kwargs):
    return rapidjson.dumps(obj, *args, number_mode=rapidjson.NM_DECIMAL, **kwargs)


def rapidjson_loads(s, *args, **kwargs):
    return rapidjson.loads(s, *args, number_mode=rapidjson.NM_DECIMAL, **kwargs)


class Entity(BaseModel):
    value: Decimal

    class Config:
        json_loads = rapidjson_loads
        json_dumps = rapidjson_dumps
```
```python
>>> # Precision beyond what float can handle and with significant trailing zero.
>>> entity = Entity(value=Decimal("1.12345678901234567890"))
>>> # Serializes to JSON number type with full precision retained, including trailing zero.
>>> entity.json()
'{"value":1.12345678901234567890}'
>>> # Round trips back to Python Decimal with full precision retained, including trailing zero.
>>> Entity.parse_raw(entity.json())
Entity(value=Decimal('1.12345678901234567890'))
```

**Comment by jdjohnso-sketch** [comment]
tl;dr
It's a _BAD_ idea to ever encode Python's **Decimal** as a JSON **number**.
Python's **Decimal** is a binary coded decimal (BCD).
Javascript's **number** is a binary floating point number.
The two datatypes have different domains of values that can be _exactly_ represented. Note, this isn't a matter of _precision_. It's fundamental to the way the values are encoded by the languages' interpreters.
To avoid data loss, Python's **Decimal** _must_ be JSON encoded as a string (and only a JSON string) so receiving Javascript can convert it directly into **Bigint** (or similar BCD implementation) without going through **number** first.


Explaination:
I'm not saying anything here that hasn't already been said in some of the other comments. But since this issue has been reopened, there's clearly still some confusion. So I'll take a stab at saying it slightly differently in case that helps.

If you're only ever working in Python, something like @gdub's Python Decimal -> JSON number -> Python Decimal example _might_ work, depending on your serializer/deserializer. This is because JSON is just a text format. So if the Python serializer writes all of the digits to the JSON text, the Python deserializer might be able to read them all back. But what did it save you? The parsing time for two extra quotation marks?

However, if you ever load those JSON files into Javascript, you're very likely going to be in trouble. Javascript stores its **number** variables as arbitrarily-sized, binary floating point. There are many decimal numbers (e.g. 0.1 and 0.2) that simply cannot be represented exactly in binary floating point, no matter how much precision you have. This is why 0.1 + 0.2 != 0.3 (see https://stackoverflow.com/questions/588004/is-floating-point-math-broken). The same is true, BTW, for Python's native **float**.

If the Python data structure has gone to the trouble of storing the numbers in **Decimal** instead of **float**, it's because semantically the decimal digits of the number are important (e.g. financial transactions with US dollars and cents). Changing a deposit amount from $10.10 to $10.1000000000000000055511151231257827021181583404541015625 is unacceptable. 

Encoding a Python **Decimal** as a JSON string, at least gives the receiving Javascript code the chance to convert the string directly into a Javascript **Bigint**. The author of the receiving Javascript still might naively stuff it into a **number** and lose the original values. But at least then it's on them and not on the sender.

This is _NOT_ just an academic discussion. This causes real-world problems. I've personally worked on two different ETL projects (extract-transform-load) for major US banks where the bank's programmers got cute with their Python -> JSON serialization to force Python **Decimal** to be serialized as a JSON **number** because "it looked weird to have them as strings in the JSON file". Fortunately unit tests caught one of them before it went live. The other one was found by _many_ of the bank's customers.

**Comment by sdasun** [comment]
@jdjohnso-sketch 

I completely agree with the technical premise here: binary floating point is a poor container for precise decimal data. However, there is a distinction between 'best practice' and 'enforced constraints.'

In many real world scenarios, we are building consumers for legacy systems or third party APIs that strictly enforce a 'number' type in their schema. If I'm forced to send a string to a system that expects a float, I've broken the integration.

Pydantic should absolutely be 'safe by default' (serializing to strings), but it shouldn't be 'safe by force.' Developers need a first-class way to opt into float-serialization when the downstream contract demands it, even if we acknowledge the potential for data loss at the edges.

**Comment by peter-gerlagh-kyos** [comment]
Seeing this suggested as the solution is a little horrifying
```
FloatDecimal = Annotated[
    Decimal,
    PlainSerializer(float, return_type=float, when_used='json')
]
```
Encoding a Decimal as a json number shouldn't require you to convert it into a float first. You lose precision in the process of converting it to float. json natively stores Decimal numbers (without loss of precision) and uses them as representations of floats. Not the other way round.

I would quite like to have a solution where I can get jsons that look like `'{"x": 1.23}'` instead of  `'{"x": "1.23"}'` without running into surprises like 
```
json.dumps(json.loads("123456789112233445566778899.0")) == '1.2345678911223344e+26'
json.loads("123456789112233445566778899.0") != json.loads("123456789112233445566778899")
json.dumps(json.loads("0.123456789112233445566778899")) == '0.12345678911223344'
```

(Yes all those examples of nasty use the Python standardlib so that I can write them without defining a pydantic Model.)

This is currently possible using simplejson 4.1.1

**Comment by devnev39** [comment]
@peter-gerlagh-kyos 

I was reading up on same topic in python since I am currently working on a finance application which has it's services written in python. The services has float as data type for amounts which I was looking up to change to Decimal. And I ran into same issue of serialising Double to string.

What if I round the value while serialising and return the response. Let's say that rounding to 3rd decimal place is enough for me. Also same, while getting data from FE I will round it to 3rd decimal place. Once in backend each calculation is done in Decimal. 

Will this work for avoiding the precision errors ?

**Comment by peter-gerlagh-kyos** [comment]
@devnev39 as you can see in the code snippet I posted, going through float will cause you to lose information when there are at least 17 significant digits.
If you know that you need exactly 3 decimal places, the procedure you describe won't actually result in bugs unless the stored numbers are greater than 99 trillion euros/dollars/pounds. So in the absence of hyperinflation you're probably safe.

I still wouldn't do it that way myself.
`simplejson.dumps(your_object.model_dump())` is a simple way to get the nice json without having to worry about float accuracy.
You can also use the double-stringification approach pushed by Pydantic.

**Comment by rafalkrupinski** [comment]
It would be great to have simple ways of encoding decimals either as numbers (unquoted) or as strings (quoted).
Both are useful depending on the context and it's the user that should decide. One can be the default, the other opt-in.

Unquoted encoding should *not* go via float, tho limiting their precision to double by default would be sensible thing to do, IMO.
